models/query_class.go

```diff
@@ -0,0 +1,441 @@
+	:m_innodb_io_r_ops_min,
+	:m_innodb_io_r_ops_max,
+	:m_innodb_io_r_ops_p99,
+	:m_innodb_io_r_ops_hg,
+	:m_innodb_io_r_bytes_cnt,
+	:m_innodb_io_r_bytes_sum,
+	:m_innodb_io_r_bytes_min,
+	:m_innodb_io_r_bytes_max,
+	:m_innodb_io_r_bytes_p99,
+	:m_innodb_io_r_bytes_hg,
+	:m_innodb_io_r_wait_cnt,
+	:m_innodb_io_r_wait_sum,
+	:m_innodb_io_r_wait_min,
+	:m_innodb_io_r_wait_max,
+	:m_innodb_io_r_wait_p99,
+	:m_innodb_io_r_wait_hg,
+	:m_innodb_rec_lock_wait_cnt,
+	:m_innodb_rec_lock_wait_sum,
+	:m_innodb_rec_lock_wait_min,
+	:m_innodb_rec_lock_wait_max,
+	:m_innodb_rec_lock_wait_p99,
+	:m_innodb_rec_lock_wait_hg,
+	:m_innodb_queue_wait_cnt,
+	:m_innodb_queue_wait_sum,
+	:m_innodb_queue_wait_min,
+	:m_innodb_queue_wait_max,
+	:m_innodb_queue_wait_p99,
+	:m_innodb_queue_wait_hg,
+	:m_innodb_pages_distinct_cnt,
+	:m_innodb_pages_distinct_sum,
+	:m_innodb_pages_distinct_min,
+	:m_innodb_pages_distinct_max,
+	:m_innodb_pages_distinct_p99,
+	:m_innodb_pages_distinct_hg,
+	:m_query_length_cnt,
+	:m_query_length_sum,
+	:m_query_length_min,
+	:m_query_length_max,
+	:m_query_length_p99,
+	:m_query_length_hg,
+	:m_bytes_sent_cnt,
+	:m_bytes_sent_sum,
+	:m_bytes_sent_min,
+	:m_bytes_sent_max,
+	:m_bytes_sent_p99,
+	:m_bytes_sent_hg,
+	:m_tmp_tables_cnt,
+	:m_tmp_tables_sum,
+	:m_tmp_tables_min,
+	:m_tmp_tables_max,
+	:m_tmp_tables_p99,
+	:m_tmp_tables_hg,
+	:m_tmp_disk_tables_cnt,
+	:m_tmp_disk_tables_sum,
+	:m_tmp_disk_tables_min,
+	:m_tmp_disk_tables_max,
+	:m_tmp_disk_tables_p99,
+	:m_tmp_disk_tables_hg,
+	:m_tmp_table_sizes_cnt,
+	:m_tmp_table_sizes_sum,
+	:m_tmp_table_sizes_min,
+	:m_tmp_table_sizes_max,
+	:m_tmp_table_sizes_p99,
+	:m_tmp_table_sizes_hg,
+	:m_qc_hit_sum,
+	:m_full_scan_sum,
+	:m_full_join_sum,
+	:m_tmp_table_sum,
+	:m_tmp_table_on_disk_sum,
+	:m_filesort_sum,
+	:m_filesort_on_disk_sum,
+	:m_select_full_range_join_sum,
+	:m_select_range_sum,
+	:m_select_range_check_sum,
+	:m_sort_range_sum,
+	:m_sort_rows_sum,
+	:m_sort_scan_sum,
+	:m_no_index_used_sum,
+	:m_no_good_index_used_sum,
+	:grpstr,
+	:grpint,
+	:labint_key
+	:labint_value
+  )
+`
+
+type QueryClass struct {
+	db *sqlx.DB
+}
+
+func NewQueryClass(db *sqlx.DB) QueryClass {
+	return QueryClass{db: db}
+}
+
+func (qc *QueryClass) Save(agentMsg *collectorpb.AgentMessage) error {
+	type QueryClass struct {
+		PeriodStart   time.Time `json:"period_start_ts"`
+		ExampleType   string    `json:"example_type_s"`
+		ExampleFormat string    `json:"example_format_s"`
+		LabelsKey     []string  `json:"labels_key"`
+		LabelsValues  []string  `json:"labels_value"`
+		WarningsCode  []string  `json:"warnings_code"`
+		WarningsCount []uint64  `json:"warnings_count"`
+		ErrorsCode    []string  `json:"errors_code"`
+		ErrorsCount   []uint64  `json:"errors_count"`
+		LabintKey     []uint32  `json:"labint_key"`
+		LabintValue   []uint32  `json:"labint_value"`
+		*collectorpb.QueryClass
+	}
+
+	qc.db.Mapper = reflectx.NewMapperTagFunc("json", strings.ToUpper, func(value string) string {
```

Comment:
I'm pretty sure that would be the slowest place… I guess that's okay for now, but something we should definitely look into before 2.0 release.